### ISSUE TITLE
On Windows 10, Node.js 4.3.x, always gets ETIMEDOUT

### ISSUE THREAD

**Issue by truedat101** [box/box-node-sdk#10]
Using example from:
https://github.com/box/box-node-sdk/blob/master/docs/files.md#upload-a-file

My simple upload function below will always fail with ETIMEDOUT.  Can you clarify the reason for this failure?  I passed authentication steps, even created a folder.  This error seems to happen in the core, not in the API.  Is it a problem with auth?  Windows issue?

```
var stream = fs.createReadStream(path.normalize('./public/img/icons.png'));
client.files.uploadFile(folder_id, 'icons.png', stream, function(err, data){
  if (err) {
    console.log("Error sending a zipfile into " + folder_id + " to Box, Reason: " + err);
  } else {
   console.log("Zip sent to Box");
}
});
```

**Comment by truedat101** [comment]
Tried this on linux, and fails as well.  Oddly, the calls I do leading up to this all succeed:

- sdk.getTokensRefreshGrant
- client.folders.create
- client.users.get

The files I've tried uploading are my very own README.md (7K), icon ping (about 100K).  Always times out.   Are there some params I need to tune on config?

**Comment by yukoyuko** [comment]
I've also been able to reproduce on Windows 8. I use Node.js 4.4.7.

**Comment by djordan** [comment]
@truedat101 & @yukoyuko, thanks for letting us know. We'll get back to you once we get a chance to look at it.

**Comment by yukoyuko** [comment]
Thank you!

**Comment by truedat101** [comment]
Given the lack of bugs filed, either it's perfect code or just not a critical mass using it yet.  I find it odd the fact other functions work so well, but this function, with a different URL behind it fails 100% for whatever folder and file used.

**Comment by truedat101** [comment]
Any tips on the debugging of the api timeout would be appreciated as a starting point.  I am currently using defaults for config, but did also attempt to change timeout values.

**Comment by djordan** [comment]
@truedat101, have you tried doing an upload with the example app? Trying that would help us  figure out if we're looking at a code issue or an environment issue.

**Comment by mattwiller** [comment]
I'm able to reproduce on Node v4.4.2 on OSX, but it depends on the time between creating the read stream and making the upload request.  When using the node REPL to test, if I do

```
> var stream = fs.createReadStream('./README.md'); client.files.uploadFile('0', 'readme.txt', stream, (err, data) => console.log(err, data));
```

I get the error, but if I enter the statements separately (and thus have a couple seconds between them), it works fine.  How strange.  I've tried using the built-in debugging statements from the `request` module, but they don't have anything helpful in the error case.  I'll dig in over the next few days and set up a test case to try and find out what's going on.

**Comment by truedat101** [comment]
@djordan Actually no just went straight in for my specific use case which was pretty simple, make a directory (that worked) and then upload a file.  I'll go back and try the example if that can establish a baseline.

**Comment by djordan** [comment]
@truedat101 - thanks for helping out. Since @mattwiller was able to reproduce it, I think you can hold off on trying the sample app.

**Comment by mattwiller** [comment]
I was able to root-cause the issue and identify a possible workaround.  The problem is at https://github.com/box/box-node-sdk/blob/master/lib/api-request-manager.js#L44, where we merge the specific API request options (including a reference to the stream) with the SDK defaults.  The `lodash.merge` library that we use to merge the config options appears to be creating a deep copy of the stream object.  This is a problem because usually the original stream object is asynchronously opening the file in the background at the time when it's copied.  When the file is opened for reading, the original stream object gets updated and has data flowing but the copy does not.  So, when we're trying to make the request and read data from the copied stream, the request ends up hanging waiting for the data (which will never arrive). Eventually, the request times out before completing.

I'll work on implementing a fix, but for the meanwhile I was able to work around the problem by manually opening the file descriptor and then creating the stream with it.  Modifying the example from @truedat101 above:

```js
fs.open(path.normalize('./public/img/icons.png'), 'r', function(err, fd) {

    var stream = fs.createReadStream(path.normalize('./public/img/icons.png'), {fd: fd});
    client.files.uploadFile(folder_id, 'icons.png', stream, function(err, data){
        if (err) {
            console.log("Error sending a zipfile into " + folder_id + " to Box, Reason: " + err);
        } else {
            console.log("Zip sent to Box");
        }
    });
});
```

Could you give that a try and let me know if it works?

**Comment by manfi** [comment]
Hi, I found the same problem referenced by this issue trying to stream from s3 to box
``` javascript
var stream = s3.getObject(params).createReadStream();
client.files.uploadFile(playlistFolder, fileName, stream, function (err, data) {...}
```
Then I tried the work around with an example file and got 500 response from the server, then I tried using a buffer with the same result
```javascript
 var stream1 = require('fs').createReadStream('package.json', {fd: fd});
  stream1.on('data', function (data) {
      content.push(data);
  })
  stream1.on('end', function (data) {
    client.files.uploadFile(playlistFolder, "package.json", Buffer.concat(content), function (err, data) {
      if (err) {
        reject({
          code: 500,
          message: "unable to upload",
          error: err
        });
      } else {
        resolve({code: 200, data});
      }
    });
  })
```
Tested on OSX 10.11 and have also attached the request a response details.
Any idea on what is causing the problem ?

Thanks

[Request_to_box.txt](https://github.com/box/box-node-sdk/files/486859/Request_to_box.txt)

**Comment by manfi** [comment]
Ok I was passing the wrong parameter in playlistfolder, the work around works

**Comment by truedat101** [comment]
Glad the work around is working, but wondering when the fix will be released?

**Comment by mattwiller** [comment]
@truedat101 The fix for this issue was merged today; we plan to cut a new release of the SDK by mid-week.

**Comment by truedat101** [comment]
Great, thanks for sorting this out.  Normally we don't use Windows for Node.js stuff, but happens to be needed for the current project.

**Comment by mattwiller** [comment]
@truedat101 @yukoyuko @manfi To close the loop on this, we just released v1.1.0 of the SDK, which should fix this issue.  Please let us know if you're still experiencing difficulties with the new version!

**Comment by yukoyuko** [comment]
@mattwiller With the latest version of SDK, I am now able to upload files no problem. Thank you very much!
